app/javascript/components/student/editor/FileEditor.tsx

```diff
@@ -3,28 +3,71 @@ import React, {
   useImperativeHandle,
   useRef,
   useCallback,
+  useState,
 } from 'react'
 import { File } from '../Editor'
 import MonacoEditor from 'react-monaco-editor'
 import * as monacoEditor from 'monaco-editor/esm/vs/editor/editor.api'
+import { useLocalStorage } from '../../../utils/use-local-storage'
 
 export type FileEditorHandle = {
   getFile: () => File
 }
 
 type FileEditorProps = {
   file: File
+  syntaxHighlighter: string
+  onRunTests: () => void
 }
 
 export const FileEditor = forwardRef<FileEditorHandle, FileEditorProps>(
-  ({ file }, ref) => {
+  ({ file, syntaxHighlighter, onRunTests }, ref) => {
+    const [theme, setTheme] = useState('vs')
+    const [options, setOptions] = useState({
+      minimap: { enabled: false },
+      wordWrap: 'on',
+    })
+    const [content, setContent] = useLocalStorage(
+      `${file.name}-editor-content`,
+      file.content
+    )
     const editorRef = useRef<monacoEditor.editor.IStandaloneCodeEditor>()
     const editorDidMount = useCallback(
       (editor) => {
+        editor.addAction({
+          id: 'runTests',
+          label: 'Run tests',
+          keybindings: [monaco.KeyCode.F2],
+          run: onRunTests,
+        })
         editorRef.current = editor
       },
       [editorRef]
     )
+    const editorChanged = useCallback(
+      (newValue, e) => {
+        setContent(newValue)
```

Comment:
This is fine for now, but the issue with this is that the editor has changed, and you're instantly writing to storage, which triggers a re-render. It's probably not what you want!

My recommendation would be to periodically store directly, at least debounce this. On blur, store definitely.

Reply:
I've set an interval to save the code to storage every 1 second.